hudi-utilities/src/test/java/org/apache/hudi/utilities/deltastreamer/TestTransformer.java

```diff
@@ -92,4 +211,95 @@ public Dataset<Row> apply(JavaSparkContext jsc, SparkSession sparkSession, Datas
       return rowDataset.withColumn("timestamp", functions.col("timestamp").multiply(multiplier).plus(increment));
     }
   }
+
+  /**
+   * Provides a transformedSchema implementation for FlatteningTransformer.
+   */
+  public static class FlatteningTransformerWithTransformedSchema extends FlatteningTransformer {
+
+    @Override
+    public Option<StructType> transformedSchema(JavaSparkContext jsc, SparkSession sparkSession, StructType incomingStruct, TypedProperties properties) {
+      String flattenedSelect = flattenSchema(incomingStruct, null);
+      String[] cols = flattenedSelect.split(",");
+      List<Pair<String, String>> replacements = new LinkedList<>();
+      for (String col : cols) {
+        String[] names = col.split(" as ");
+        if (!names[0].equals(names[1])) {
+          replacements.add(Pair.of(names[0], names[1]));
+        }
+      }
+
+      List<StructField> incomingFields = Arrays.asList(incomingStruct.fields());
+      List<StructField> transformedFields = new LinkedList<>(incomingFields);
+      Set<StructField> fieldsToRemove = new HashSet<>();
+      for (Pair<String, String> replacement : replacements) {
+        String fieldToRemoveName = replacement.getKey().replaceAll("\\..*", "");
+        StructField fieldToAdd = StructUtils.getField(incomingStruct, replacement.getKey()).get();
+        StructField fieldToRemove = transformedFields.stream().filter(f -> f.name().equals(fieldToRemoveName)).findAny().get();
+        fieldsToRemove.add(fieldToRemove);
+        transformedFields.add(transformedFields.indexOf(fieldToRemove), new StructField(replacement.getKey().replaceAll("\\.", "_"),
+            fieldToAdd.dataType(), fieldToAdd.nullable(), fieldToAdd.metadata()));
+      }
+      transformedFields.removeAll(fieldsToRemove);
+
+      return Option.of(new StructType(transformedFields.toArray(new StructField[0])));
+    }
+  }
+
+  /**
+   * Does not provide transformedSchema implementation for FlatteningTransformer.
+   */
+  public static class FlatteningTransformerWithoutTransformedSchema extends FlatteningTransformer {
+    @Override
+    public Option<StructType> transformedSchema(JavaSparkContext jsc, SparkSession sparkSession, StructType incomingStruct, TypedProperties properties) {
+      return Option.empty();
+    }
+  }
+
+  /**
+   * Adds a new column named random in the dataset.
+   */
+  public static class AddColumnTransformer implements Transformer {
```

Comment:
Let's make a transform that operates on a field that does not actually exist in the source to make sure that we are throwing the proper exception in this case. Something like:
`rowDataset.withColumn("random", functions.lit(5).multiply(functions.col("unknown_column")));`

Reply:
Addressed